File: apps/api/src/tutor_api/agent/capability.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import secrets
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from pathlib import Path
from typing import Any
from uuid import UUID

from sqlalchemy import select
from sqlalchemy.orm import Session

from tutor_api.classrooms.models import Classroom, ClassroomMembership, ClassroomRole
from tutor_api.identity.models import User
from tutor_api.knowledge.access import list_readable_knowledge_bases
from tutor_api.knowledge.models import KnowledgeBase
from tutor_api.spaces.models import Space, SpaceKind


class CapabilityError(ValueError):
    """A stable, non-secret capability validation failure."""

    def __init__(self, code: str) -> None:
        super().__init__(code)
        self.code = code


@dataclass(frozen=True)
class WorkspaceGrant:
    knowledge_base_id: UUID
    vault_root: str
    actions: tuple[str, ...]


_AGENT_TOOL_CATEGORIES = ("vault", "shell", "web", "mcp", "skills", "subagents")
_AGENT_KNOWLEDGE_BASE_ACTIONS = {"read", "write", "delete"}


def _secret(value: Any) -> str:
    getter = getattr(value, "get_secret_value", None)
    return getter() if getter else str(value)


def _setting(settings: Any, name: str, default: Any = None) -> Any:
    return getattr(settings, name, default)


def _b64(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def _unb64(data: str) -> bytes:
    return base64.urlsafe_b64decode(data + "=" * (-len(data) % 4))


def _canonical(payload: dict[str, Any]) -> bytes:
    return json.dumps(payload, sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode()


def _vault_path(root: Path, space_id: UUID, knowledge_base_id: UUID) -> str:
    return str((root / "spaces" / str(space_id) / str(knowledge_base_id)).resolve())


def _can_write(session: Session, user: User, knowledge_base: KnowledgeBase) -> bool:
    space = session.get(Space, knowledge_base.space_id)
    if space is None:
        return False
    if space.kind is SpaceKind.PERSONAL:
        return space.owner_id == user.id
    role = session.scalar(
        select(ClassroomMembership.role)
        .join(Classroom, Classroom.id == ClassroomMembership.classroom_id)
        .where(
            Classroom.space_id == space.id,
            ClassroomMembership.user_id == user.id,
        )
    )
    return role in {ClassroomRole.OWNER, ClassroomRole.TEACHER}


def issue_workspace_capability(
    session: Session,
    user: User,
    *,
    session_id: UUID,
    settings: Any,
    now: datetime | None = None,
) -> str:
    issued = now or datetime.now(UTC)
    if issued.tzinfo is None:
        issued = issued.replace(tzinfo=UTC)
    ttl = int(_setting(settings, "agent_capability_ttl_seconds", 300))
    if not 30 <= ttl <= 3600:
        raise CapabilityError("capability_ttl_invalid")
    root = Path(_setting(settings, "agent_vault_root", Path.cwd() / ".agent-vault"))
    grants: list[WorkspaceGrant] = []
    for kb in list_readable_knowledge_bases(session, user):
        writable = _can_write(session, user, kb)
        grants.append(
            WorkspaceGrant(
                knowledge_base_id=kb.id,
                vault_root=_vault_path(root, kb.space_id, kb.id),
                actions=("read", "write", "delete") if writable else ("read",),
            )
        )
    payload = {
        "version": "1.0",
        "user_id": str(user.id),
        "session_id": str(session_id),
        "issued_at": issued.isoformat(),
        "expires_at": (issued + timedelta(seconds=ttl)).isoformat(),
        "nonce": secrets.token_urlsafe(24),
        "grants": [
            {
                "knowledge_base_id": str(grant.knowledge_base_id),
                "actions": list(grant.actions),
            }
            for grant in grants
        ],
        "tool_categories": list(_AGENT_TOOL_CATEGORIES),
        "vault_roots": [grant.vault_root for grant in grants],
    }
    body = _b64(_canonical(payload))
    secret = _secret(_setting(settings, "agent_capability_secret"))
    if len(secret) < 32:
        raise CapabilityError("capability_secret_invalid")
    signature = _b64(hmac.new(secret.encode(), body.encode(), hashlib.sha256).digest())
    return f"{body}.{signature}"


def verify_workspace_capability(
    token: str,
    *,
    settings: Any,
    expected_session_id: UUID | None = None,
    expected_user_id: UUID | None = None,
    now: datetime | None = None,
) -> dict[str, Any]:
    try:
        body, supplied = token.split(".", 1)
        secret = _secret(_setting(settings, "agent_capability_secret"))
        expected = _b64(hmac.new(secret.encode(), body.encode(), hashlib.sha256).digest())
        if not hmac.compare_digest(supplied, expected):
            raise CapabilityError("capability_signature_invalid")
        payload = json.loads(_unb64(body))
    except CapabilityError:
        raise
    except Exception as error:
        raise CapabilityError("capability_invalid") from error
    grants = payload.get("grants")
    tool_categories = payload.get("tool_categories")
    vault_roots = payload.get("vault_roots")
    if (
        payload.get("version") != "1.0"
        or not isinstance(grants, list)
        or not isinstance(tool_categories, list)
        or not isinstance(vault_roots, list)
        or any(
            not isinstance(grant, dict)
            or not isinstance(grant.get("knowledge_base_id"), str)
            or not isinstance(grant.get("actions"), list)
            or any(action not in _AGENT_KNOWLEDGE_BASE_ACTIONS for action in grant["actions"])
            for grant in grants
        )
        or any(category not in _AGENT_TOOL_CATEGORIES for category in tool_categories)
        or any(not isinstance(root, str) or not root for root in vault_roots)
    ):
        raise CapabilityError("capability_invalid")
    current = now or datetime.now(UTC)
    expires = datetime.fromisoformat(payload["expires_at"])
    if expires.tzinfo is None:
        expires = expires.replace(tzinfo=UTC)
    if current >= expires:
        raise CapabilityError("capability_expired")
    if expected_session_id is not None and payload.get("session_id") != str(expected_session_id):
        raise CapabilityError("capability_session_mismatch")
    if expected_user_id is not None and payload.get("user_id") != str(expected_user_id):
        raise CapabilityError("capability_user_mismatch")
    return payload

File: apps/agent-runtime/src/security/capability.ts
import { createHmac, timingSafeEqual } from "node:crypto";

import type {
  AgentKnowledgeBaseAction,
  AgentKnowledgeBaseGrant,
  AgentToolCategory,
  AgentWorkspaceCapabilityPayload,
} from "@textbook-agent/agent-protocol";

const ACTIONS = new Set<AgentKnowledgeBaseAction>(["read", "write", "delete"]);
const TOOLS = new Set<AgentToolCategory>(["vault", "shell", "web", "mcp", "skills", "subagents"]);
const NONCE_PATTERN = /^[A-Za-z0-9_-]{16,128}$/;

export class CapabilityError extends Error {
  constructor(readonly code: string, message: string) {
    super(message);
    this.name = "CapabilityError";
  }
}

export class VerifiedCapability {
  constructor(readonly payload: AgentWorkspaceCapabilityPayload) {}

  requireGrant(knowledgeBaseId: string, action: AgentKnowledgeBaseAction): AgentKnowledgeBaseGrant {
    const grant = this.payload.grants.find(item => item.knowledge_base_id === knowledgeBaseId);
    if (!grant || !grant.actions.includes(action)) {
      throw new CapabilityError("capability_denied", `Capability does not grant ${action} on ${knowledgeBaseId}`);
    }
    return grant;
  }

  requireTool(category: AgentToolCategory): void {
    if (!this.payload.tool_categories.includes(category)) {
      throw new CapabilityError("capability_denied", `Capability does not grant tool category ${category}`);
    }
  }
}

export interface VerifyCapabilityOptions {
  secret: string | Buffer;
  sessionId: string;
  now?: Date;
}

export function signCapability(payload: AgentWorkspaceCapabilityPayload, secret: string | Buffer): string {
  const encoded = Buffer.from(JSON.stringify(payload), "utf8").toString("base64url");
  return `${encoded}.${signature(encoded, secret).toString("base64url")}`;
}

export function verifyCapability(token: string, options: VerifyCapabilityOptions): VerifiedCapability {
  const parts = token.split(".");
  if (parts.length !== 2 || !parts[0] || !parts[1]) fail("capability_invalid", "Malformed capability");
  const expected = signature(parts[0], options.secret);
  let provided: Buffer;
  try { provided = Buffer.from(parts[1], "base64url"); } catch { fail("capability_invalid", "Malformed signature"); }
  if (provided.length !== expected.length || !timingSafeEqual(provided, expected)) fail("capability_invalid", "Invalid capability signature");

  let payload: AgentWorkspaceCapabilityPayload;
  try { payload = JSON.parse(Buffer.from(parts[0], "base64url").toString("utf8")) as AgentWorkspaceCapabilityPayload; }
  catch { fail("capability_invalid", "Malformed capability payload"); }
  validatePayload(payload);
  if (payload.session_id !== options.sessionId) fail("capability_session_mismatch", "Capability is bound to another session");
  const now = (options.now ?? new Date()).getTime();
  const issued = Date.parse(payload.issued_at);
  const expires = Date.parse(payload.expires_at);
  if (!Number.isFinite(issued) || !Number.isFinite(expires) || issued > expires) fail("capability_invalid", "Capability timestamps are invalid");
  if (now < issued - 60_000) fail("capability_not_yet_valid", "Capability is not active yet");
  if (now >= expires) fail("capability_expired", "Capability has expired");
  return new VerifiedCapability(payload);
}

function validatePayload(payload: AgentWorkspaceCapabilityPayload): void {
  if (!payload || typeof payload !== "object") fail("capability_invalid", "Capability payload must be an object");
  for (const key of ["version", "user_id", "session_id", "issued_at", "expires_at", "nonce"] as const) {
    if (typeof payload[key] !== "string" || payload[key].length === 0) fail("capability_invalid", `Capability ${key} is required`);
  }
  if (!NONCE_PATTERN.test(payload.nonce)) fail("capability_invalid", "Capability nonce is invalid");
  if (!Array.isArray(payload.grants) || !Array.isArray(payload.tool_categories) || !Array.isArray(payload.vault_roots)) fail("capability_invalid", "Capability arrays are invalid");
  for (const grant of payload.grants) {
    if (!grant || typeof grant.knowledge_base_id !== "string" || !Array.isArray(grant.actions) || grant.actions.some((action: AgentKnowledgeBaseAction) => !ACTIONS.has(action))) fail("capability_invalid", "Capability grant is invalid");
  }
  if (payload.tool_categories.some(tool => !TOOLS.has(tool))) fail("capability_invalid", "Capability tool category is invalid");
  if (payload.vault_roots.some(root => typeof root !== "string" || root.length === 0)) fail("capability_invalid", "Capability vault root is invalid");
}

function signature(encodedPayload: string, secret: string | Buffer): Buffer {
  return createHmac("sha256", secret).update(encodedPayload, "utf8").digest();
}

function fail(code: string, message: string): never { throw new CapabilityError(code, message); }
